File: src/admin/services/discounts.py
"""Global discount management service."""

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from src.admin.schemas import (
    CreateGlobalDiscountRequest,
    GlobalDiscountListItem,
    GlobalDiscountListResponse,
    UpdateGlobalDiscountRequest,
)
from src.db.models.global_discount import GlobalDiscount


async def create_global_discount(
    session: AsyncSession,
    request: CreateGlobalDiscountRequest,
) -> GlobalDiscount:
    discount = GlobalDiscount(
        name=request.name,
        target_plan=request.target_plan,
        discount_percent=request.discount_percent,
        active_until=request.active_until,
        is_active=True,
    )
    session.add(discount)
    await session.commit()
    await session.refresh(discount)
    return discount


async def list_global_discounts(
    session: AsyncSession,
    page: int = 1,
    page_size: int = 20,
    is_active: bool | None = None,
) -> GlobalDiscountListResponse:
    query = select(GlobalDiscount)
    count_query = select(func.count()).select_from(GlobalDiscount)

    if is_active is not None:
        query = query.where(GlobalDiscount.is_active == is_active)
        count_query = count_query.where(GlobalDiscount.is_active == is_active)

    total = await session.scalar(count_query) or 0

    query = query.order_by(GlobalDiscount.created_at.desc())
    query = query.offset((page - 1) * page_size).limit(page_size)

    result = await session.execute(query)
    discounts = result.scalars().all()
    items = [GlobalDiscountListItem.model_validate(d) for d in discounts]

    return GlobalDiscountListResponse(items=items, total=total, page=page, page_size=page_size)


async def update_global_discount(
    session: AsyncSession,
    discount_id: int,
    request: UpdateGlobalDiscountRequest,
) -> bool:
    discount = await session.get(GlobalDiscount, discount_id)
    if not discount:
        return False

    if request.name is not None:
        discount.name = request.name
    if request.discount_percent is not None:
        discount.discount_percent = request.discount_percent
    if request.active_until is not None:
        discount.active_until = request.active_until
    if request.is_active is not None:
        discount.is_active = request.is_active

    await session.commit()
    return True


async def delete_global_discount(
    session: AsyncSession,
    discount_id: int,
) -> bool:
    discount = await session.get(GlobalDiscount, discount_id)
    if not discount:
        return False

    await session.delete(discount)
    await session.commit()
    return True

File: admin-frontend/src/api/endpoints/discounts.ts
import { api } from '@/api/client'

export interface GlobalDiscountListItem {
  id: number
  name: string
  target_plan: string
  discount_percent: number
  active_from: string
  active_until: string | null
  is_active: boolean
  created_at: string
}

export interface GlobalDiscountListResponse {
  items: GlobalDiscountListItem[]
  total: number
  page: number
  page_size: number
}

export interface CreateGlobalDiscountRequest {
  name: string
  target_plan: 'monthly' | 'yearly' | 'all'
  discount_percent: number
  active_until?: string
}

export async function getGlobalDiscounts(
  params: Record<string, unknown> = {},
): Promise<GlobalDiscountListResponse> {
  const { data } = await api.get<GlobalDiscountListResponse>('/global-discounts', { params })
  return data
}

export async function createGlobalDiscount(
  request: CreateGlobalDiscountRequest,
): Promise<GlobalDiscountListItem> {
  const { data } = await api.post<GlobalDiscountListItem>('/global-discounts', request)
  return data
}

export async function updateGlobalDiscount(
  id: number,
  request: Partial<CreateGlobalDiscountRequest & { is_active: boolean }>,
): Promise<void> {
  await api.patch(`/global-discounts/${id}`, request)
}

export async function deleteGlobalDiscount(id: number): Promise<void> {
  await api.delete(`/global-discounts/${id}`)
}
